README
NAME
    MooX::NewDefaults - Alter attribute defaults with less pain

VERSION
    version 0.001002

SYNOPSIS
    Concept and documentation liberally stolen from "MooseX::NewDefaults"

        package One;
        use Moo;

        has A => (is => 'ro', default => sub { 'say ahhh' });
        has B => (is => 'ro', default => sub { 'say whoo' });

        package Two;
        use Moo;
        use MooX::NewDefaults;

        extends 'One';

        # sugar for defining a new default
        default_for A => sub { 'say oooh' };

        # this is also legal
        default_for B => 'say oooh';

DESCRIPTION
    Ever start using a package from the CPAN, only to discover that it
    requires lots of subclassing and "has '+foo' => (default => ...)"? It's
    not recommended Moose best practice, and it's certainly not the
    prettiest thing ever, either.

    That's where we come in.

    This package introduces new sugar that you can use in your class,
    "default_for" (as seen above).

    e.g.

        has '+foo' => (default => sub { 'a b c' });

    ...is the same as:

        default_for foo => sub { 'a b c' };

NEW SUGAR
  default_for
    default_for() is a shortcut to extend an attribute to give it a new
    default; this default value may be any legal value for default options.

        # attribute bar defined elsewhere (e.g. superclass)
        default_for bar => 'new default';

    ... is the same as:

        has '+bar' => (default => 'new default');

AUTHORS
    *   Kent Fredric <kentnl@cpan.org>

    *   Chris Weyl <cweyl@alumni.drew.edu>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2011 by Chris Weyl
    <cweyl@alumni.drew.edu>.

    This is free software, licensed under:

      The GNU Lesser General Public License, Version 2.1, February 1999



Changes
Revision history for MooX-NewDefaults

0.001003 2017-03-09T17:40:41Z 4c95231

0.001002 2017-03-09T17:39:55Z c066958
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Make tests more compatible with older Test::More

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1
 - test: +1 ↓1 (recommends: +1 ↑1 ↓1)

0.001001 2014-08-16T17:59:02Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests updated.
 - metadata updated.

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 ↑1 (suggests: ↑2)
 - test: +1 -3 (recommends: ↑1 -1)

 [Tests]
 - update Test::ReportPrereqs Tests

0.001000 2014-06-06T15:02:57Z
 - Initial release
 - Code inspired / API compatible with MooseX::NewDefaults
 - Docs and some tests stolen there.
 - Copyright assigned there too because #yolo


lib/MooX/NewDefaults.pm
use 5.006;
use strict;
use warnings;

package MooX::NewDefaults;

our $VERSION = '0.001003';

# ABSTRACT: Alter attribute defaults with less pain

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Sub::Exporter::Progressive -setup => {
  exports => [qw( default_for )],
  groups  => {
    default => [qw( default_for )],
  },
};

sub default_for {
  my ( $name_proto, @args ) = @_;
  my $target = caller;

  my (@name_proto) = 'ARRAY' eq ref $name_proto ? @{$name_proto} : $name_proto;

  if ( @args != 1 ) {
    require Carp;
    Carp::croak(
      sprintf q[Invalid options for %s default: Single argument expected, got %s],
      join( ', ', map { "'$_'" } @name_proto ),
      scalar @args,
    );
  }
  my $coderef;
  if ( not $coderef = $target->can('has') ) {
    require Carp;
    Carp::croak( sprintf q[Calling class %s cannot "has". Did you forget to "use Moo"?], $target, );
  }

  # Calling '->has()' directly of course doesn't work, because it doesn't expect
  # $_[0] to be a class, but the attribute name.
  #
  # The class itself is baked into $target::has during `use Moo`
  return $coderef->( [ map { "+$_" } @name_proto ], default => @args );
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooX::NewDefaults - Alter attribute defaults with less pain

=head1 VERSION

version 0.001003

=head1 SYNOPSIS

Concept and documentation liberally stolen from L<< C<MooseX::NewDefaults>|MooseX::NewDefaults >>

    package One;
    use Moo;

    has A => (is => 'ro', default => sub { 'say ahhh' });
    has B => (is => 'ro', default => sub { 'say whoo' });

    package Two;
    use Moo;
    use MooX::NewDefaults;

    extends 'One';

    # sugar for defining a new default
    default_for A => sub { 'say oooh' };

    # this is also legal
    default_for B => 'say oooh';

=head1 DESCRIPTION

Ever start using a package from the CPAN, only to discover that it requires
lots of subclassing and C<< has '+foo' => (default => ...) >>?  It's not
recommended Moose best practice, and it's certainly not the prettiest thing
ever, either.

That's where we come in.

This package introduces new sugar that you can use in your class,
C<< default_for >> (as seen above).

e.g.

    has '+foo' => (default => sub { 'a b c' });

...is the same as:

    default_for foo => sub { 'a b c' };

=head1 NEW SUGAR

=head2 default_for

default_for() is a shortcut to extend an attribute to give it a new default;
this default value may be any legal value for default options.

    # attribute bar defined elsewhere (e.g. superclass)
    default_for bar => 'new default';

... is the same as:

    has '+bar' => (default => 'new default');

=head1 AUTHORS

=over 4

=item *

Kent Fredric <kentnl@cpan.org>

=item *

Chris Weyl <cweyl@alumni.drew.edu>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by Chris Weyl <cweyl@alumni.drew.edu>.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Alter attribute defaults with less pain",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>, Chris Weyl <cweyl\@alumni.drew.edu>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MooX-NewDefaults",
  "LICENSE" => "lgpl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "MooX::NewDefaults",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Sub::Exporter::Progressive" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Moo" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
